drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1015 @@
+		return spi_write_dt(&driver_cfg->spi, &tx_set);
+	}
+
+	return spi_transceive_dt(&driver_cfg->spi, &tx_set, &rx_set);
+}
+
+#define spi_nand_cmd_read(dev, opcode, dest, length) \
+	spi_nand_access(dev, opcode, 0, 0, dest, length)
+#define spi_nand_cmd_addr_read(dev, opcode, addr, dest, length) \
+	spi_nand_access(dev, opcode, NAND_ACCESS_ADDRESSED, addr, dest, length)
+#define spi_nand_cmd_write(dev, opcode) \
+	spi_nand_access(dev, opcode, NAND_ACCESS_WRITE, 0, NULL, 0)
+#define spi_nand_cmd_addr_write(dev, opcode, addr, src, length) \
+	spi_nand_access(dev, opcode, NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED, \
+		       addr, (void *)src, length)
+
+
+
+/* Everything necessary to acquire owning access to the device.
+ *
+ * This means taking the lock .
+ */
+static void acquire_device(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct spi_nand_data *const driver_data = dev->data;
+
+		k_sem_take(&driver_data->sem, K_FOREVER);
+	}
+}
+
+/* Everything necessary to release access to the device.
+ *
+ * This means releasing the lock.
+ */
+static void release_device(const struct device *dev)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		struct spi_nand_data *const driver_data = dev->data;
+
+		k_sem_give(&driver_data->sem);
+	}
+}
+
+
+static int spi_nand_get_feature(const struct device *dev,
+				uint8_t feature_addr, uint8_t *val)
+{
+        int ret = spi_nand_access(dev, SPI_NAND_CMD_GET_FEATURE,
+			NAND_ACCESS_ADDRESSED | NAND_ACCESS_8BIT_ADDR,
+			feature_addr, val, sizeof(*val));
+
+	return ret;
+}
+
+static int spi_nand_set_feature(const struct device *dev,
+				uint8_t feature_addr, uint8_t val)
+{
+        int ret = spi_nand_access(dev, SPI_NAND_CMD_SET_FEATURE,
+			NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED | NAND_ACCESS_8BIT_ADDR,
+			feature_addr, &val, sizeof(val));
+
+	return ret;
+}
+
+/**
+ * @brief Wait until the flash is ready
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * This function should be invoked after every ERASE, PROGRAM, or
+ * WRITE_STATUS operation before continuing.  This allows us to assume
+ * that the device is ready to accept new commands at any other point
+ * in the code.
+ *
+ * @param dev The device structure
+ * @return 0 on success, negative errno code otherwise
+ */
+static int spi_nand_wait_until_ready(const struct device *dev)
+{
+	int ret;
+	uint8_t reg = 0;
+
+	do {
+		ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_STATUS, &reg);
+	} while (!ret && (reg & SPI_NAND_WIP_BIT));
+
+	return ret;
+}
+
+static int spi_nand_read_recovery_mode_set(const struct device *dev, uint8_t read_recovery_mode)
+{
+	int ret;
+
+	acquire_device(dev);
+
+	ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_RECOVERY, read_recovery_mode);
+	if (ret != 0) {
+		LOG_ERR("set feature failed: %d", ret);
+		return ret;
+	}
+
+	release_device(dev);
+	return ret;
+}
+
+static int spi_nand_conti_read_enable(const struct device *dev)
```

Comment:
In this function, and several others, you acquire the device at the beginning, and release at the end. However, there are many places in the function where you return in case of error without releasing the device, leading to a potential deadlock.
I suggest to replace the 'return ret' line with 'goto out', and add an 'out:' label before release_device() call.

Reply:
I have fix it.
Thanks for your advice.